src/source-windivert.c

```diff
@@ -0,0 +1,682 @@
+TmEcode DecodeWinDivertThreadDeinit(ThreadVars *, void *);
+
+/* internal helper functions */
+static TmEcode WinDivertRecvHelper(ThreadVars *tv, WinDivertThreadVars *);
+static TmEcode WinDivertVerdictHelper(ThreadVars *tv, Packet *p);
+static TmEcode WinDivertCloseHelper(WinDivertThreadVars *);
+
+void TmModuleReceiveWinDivertRegister(void)
+{
+    SCMutexInit(&g_wd_init_lock, NULL);
+
+    TmModule *tm_ptr = &tmm_modules[TMM_RECEIVEWINDIVERT];
+    memset(tm_ptr, 0, sizeof(TmModule));
+
+    tm_ptr->name = "ReceiveWinDivert";
+    tm_ptr->ThreadInit = ReceiveWinDivertThreadInit;
+    tm_ptr->PktAcqLoop = ReceiveWinDivertLoop;
+    tm_ptr->ThreadExitPrintStats = ReceiveWinDivertThreadExitStats;
+    tm_ptr->ThreadDeinit = ReceiveWinDivertThreadDeinit;
+    tm_ptr->flags = TM_FLAG_RECEIVE_TM;
+}
+
+void TmModuleVerdictWinDivertRegister(void)
+{
+    TmModule *tm_ptr = &tmm_modules[TMM_VERDICTWINDIVERT];
+    memset(tm_ptr, 0, sizeof(TmModule));
+
+    tm_ptr->name = "VerdictWinDivert";
+    tm_ptr->ThreadInit = VerdictWinDivertThreadInit;
+    tm_ptr->Func = VerdictWinDivert;
+    tm_ptr->ThreadDeinit = VerdictWinDivertThreadDeinit;
+}
+
+void TmModuleDecodeWinDivertRegister(void)
+{
+    TmModule *tm_ptr = &tmm_modules[TMM_DECODEWINDIVERT];
+    memset(tm_ptr, 0, sizeof(TmModule));
+
+    tm_ptr->name = "DecodeWinDivert";
+    tm_ptr->ThreadInit = DecodeWinDivertThreadInit;
+    tm_ptr->Func = DecodeWinDivert;
+    tm_ptr->ThreadDeinit = DecodeWinDivertThreadDeinit;
+    tm_ptr->flags = TM_FLAG_DECODE_TM;
+}
+
+/**
+ * \brief Main WinDivert packet receive pump
+ */
+TmEcode ReceiveWinDivertLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+
+    WinDivertThreadVars *wd_tv = (WinDivertThreadVars *)data;
+    wd_tv->slot = ((TmSlot *)slot)->slot_next;
+
+    while (true) {
+        if (suricata_ctl_flags & SURICATA_STOP) {
+            SCReturnInt(TM_ECODE_OK);
+        }
+
+        if (unlikely(WinDivertRecvHelper(tv, wd_tv) != TM_ECODE_OK)) {
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+
+        StatsSyncCountersIfSignalled(tv);
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode WinDivertRecvHelper(ThreadVars *tv, WinDivertThreadVars *wd_tv)
+{
+    SCEnter();
+
+#ifdef COUNTERS
+    WinDivertQueueVars *wd_qv = WinDivertGetQueue(wd_tv->thread_num);
+#endif /* COUNTERS */
+
+    /* make sure we have at least one packet in the packet pool, to prevent us
+     * from alloc'ing packets at line rate
+     */
+    PacketPoolWait();
+
+    /* obtain a packet buffer */
+    Packet *p = PacketGetFromQueueOrAlloc();
+    if (unlikely(p == NULL)) {
+        SCLogDebug(
+                "PacketGetFromQueueOrAlloc() - failed to obtain Packet buffer");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    PKT_SET_SRC(p, PKT_SRC_WIRE);
+
+    /* WinDivert needs to be fed a buffer, so we must make one available. It is
+     * highly likely we'll encounter segmentation offload so we'll just give our
```

Comment:
segmentation offload will lead to detection issues with certain rules using dsize, and these are meant to look at the original packets.

Reply:
I'm seeing this in `source-pcap.c:344-348` - namely the global option to disable offloading, and the `DisableIfaceOffloading function` (of which are only supported on POSIX OS's with `ethtool` from what I can tell):

```
if (LiveGetOffload() == 0) {
    (void)GetIfaceOffloading((char *)pcapconfig->iface, 1, 1);
} else {
    DisableIfaceOffloading(ptv->livedev, 1, 1);
}
```

Perhaps I should look into a solution for disabling offload on Windows? Or at least make the suggestion in the docs?

Reply:
Probably fine to add this to the docs. Perhaps a bit more than just saying 'disable offloading'. I wouldn't know how to do, although maybe it's an easy google.

Reply:
[This document](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/using-registry-values-to-enable-and-disable-task-offloading) may come in handy but I need to understand if restarting devices is enough or if it requires a full Windows restart.